packages/core/core/src/services/document-service/entries.ts

```diff
@@ -62,20 +68,51 @@ const createEntriesService = (uid: UID.ContentType) => {
     );
     // Component handling
     const componentData = await components.updateComponents(uid, entryToUpdate, validData as any);
-    const dataWithComponents = components.assignComponentData(validData, componentData, {
+    const dataWithComponents = components.assignComponentData(
       contentType,
-    });
-    const entryData = applyTransforms(dataWithComponents, { contentType });
+      componentData,
+      validData
+    );
+
+    const entryData = applyTransforms(contentType, dataWithComponents);
 
     return strapi.db
       .query(uid)
       .update({ ...query, where: { id: entryToUpdate.id }, data: entryData });
   }
 
+  async function publishEntry(entry: any, params = {} as any) {
+    return async.pipe(
+      omit('id'),
+      assoc('publishedAt', new Date()),
+      (draft) => {
+        const opts = { uid, locale: draft.locale, status: 'published', allowMissingId: true };
+        return transformData(draft, opts);
+      },
+      // Create the published entry
+      (draft) => createEntry({ ...params, data: draft, locale: draft.locale, status: 'published' })
+    )(entry);
+  }
+
+  async function discardDraftEntry(entry: any, params = {} as any) {
+    return async.pipe(
+      omit('id'),
+      assoc('publishedAt', null),
+      (entry) => {
+        const opts = { uid, locale: entry.locale, status: 'draft', allowMissingId: true };
+        return transformData(entry, opts);
+      },
+      // Create the draft entry
+      (data) => createEntry({ ...params, locale: data.locale, data, status: 'draft' })
+    )(entry);
+  }
+
   return {
     create: createEntry,
     delete: deleteEntry,
     update: updateEntry,
+    publish: publishEntry,
+    discardDraft: discardDraftEntry,
```

Comment:
do you have plans to add an unpublish entry method?

Reply:
it was too simple but if/when I expose them I would add it probably